### PR TITLE
amazon-image: remove some legacy code, fetch ec2 metadata in stage-2

### PR DESCRIPTION
**Pull request by lheckemann**
###### Description of changes

This should make the initramfs in EC2 images smaller, and boot faster.

###### Things done

- [x] Booted `amazonImage.x86_64` on EC2 and checked that both the specified SSH key and the user data were used correctly
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/virtualisation/amazon-image.nix [hunk 1]
```diff
@@ -64,20 +53,17 @@ in
     boot.extraModulePackages = [
       config.boot.kernelPackages.ena
     ];
-    boot.initrd.kernelModules = [ "xen-blkfront" "xen-netfront" ];
-    boot.initrd.availableKernelModules = [ "ixgbevf" "ena" "nvme" ];
```
Comment (cole-h) [inline nixos/modules/virtualisation/amazon-image.nix:68]: Does dropping `ena` from here affect the assertion on line 29? (`"ENA driver fails to build..."`)

File: nixos/modules/virtualisation/amazon-image.nix [hunk 2]
```diff
@@ -87,67 +73,14 @@ in
...
-    boot.initrd.extraUtilsCommands =
-      ''
-        # We need swapon in the initrd.
-        copy_bin_and_libs ${pkgs.util-linux}/sbin/swapon
-      '';
+    systemd.services.fetch-ec2-metadata = {
+      wantedBy = [ "multi-user.target" ];
+      after = ["network-online.target"];
+      path = [ pkgs.curl ];
+      script = builtins.readFile ./ec2-metadata-fetcher.sh;
```
Comment (cole-h) [inline nixos/modules/virtualisation/amazon-image.nix:80]: Is there a reason we need to use `readFile` instead of just passing the raw path?

Comment (lheckemann) [reply nixos/modules/virtualisation/amazon-image.nix:80]: `script` takes bash code. `"source ${./ec2-metadata-fetcher.sh}"` could work but would be awkward, and using `ExecStart` would require extra permission bits and such. Using `readFile` seemed like the simplest solution to me here.

Comment (cole-h) [reply nixos/modules/virtualisation/amazon-image.nix:80]: Would it be acceptable to `pkgs.writeShellScript` it (`pkgs.writeShellScript "..." ./ec2-metadata-fetcher.sh`) and interpolate that / use that as the ExecStart?

It's not a strong feeling, but using `builtins.readFile` here irks me for some reason.

Comment (lheckemann) [reply nixos/modules/virtualisation/amazon-image.nix:80]: That also won't work, since it expands to the path and not the file's contents.

**Comment by lheckemann** [comment]
Branch-off has taken place. I've rebased and moved the release notes section to 23.05.

**Review by cole-h — APPROVED** [review]
Looks fine to me (once the release notes conflicts have been addressed).

**Comment by lheckemann** [comment]
Rebased to fix conflict and squash in the fixup commit.

**Pull request merged**